logger.go

```diff
@@ -1,6 +1,222 @@
+// writer and a Raft formatter.
+func NewRaftLoggerWithWriter(w io.Writer, level int) log.Logger {
+	logger := log.NewLogger(w, "raft")
+	return setLevelFormatter(logger, level, createRaftFormatter("", false))
+}
+
+// NewRaftLoggerForTesting creates a new logger with the specified tag, for use
+// in unit tests.
+func NewRaftLoggerForTesting(w io.Writer, tag string) log.Logger {
+	logger := log.NewLogger(w, "raft")
+	level := log.LevelDebug
+	return setLevelFormatter(logger, level, createRaftFormatter(tag, true))
+}
+
+// Sets the level and formatter on the log, which must be a DefaultLogger.
+func setLevelFormatter(logger log.Logger, level int, formatter log.Formatter) log.Logger {
+	logger.(*log.DefaultLogger).SetLevel(level)
+	logger.(*log.DefaultLogger).SetFormatter(formatter)
+	return logger
+}
+
+// createRaftFormatter builds a formatter and looks via environment variables
+// for style customization.
+func createRaftFormatter(tag string, withLine bool) log.Formatter {
+	ret := &raftFormatter{
+		Mutex:    &sync.Mutex{},
+		tag:      tag,
+		withLine: withLine,
+	}
+	logFormat := os.Getenv("RAFT_LOG_FORMAT")
+	if logFormat == "" {
+		logFormat = os.Getenv("LOGXI_FORMAT")
+	}
+	switch strings.ToLower(logFormat) {
+	case "json", "raft_json", "raft-json", "raftjson":
+		ret.style = styleJSON
+	default:
+		ret.style = styleDefault
+	}
+	return ret
+}
+
+// raftFormatter is a thread-safe formatter.
+type raftFormatter struct {
+	*sync.Mutex
+	style    int
+	module   string
+	tag      string
+	withLine bool
+}
+
+// Format preps a log message for the logger.
+func (v *raftFormatter) Format(writer io.Writer, level int, msg string, args []interface{}) {
+	v.Lock()
+	defer v.Unlock()
+	switch v.style {
+	case styleJSON:
+		v.formatJSON(writer, level, msg, args)
+	default:
+		v.formatDefault(writer, level, msg, args)
+	}
+}
+
+// formatDefault is the default text-based log format.
+func (v *raftFormatter) formatDefault(writer io.Writer, level int, msg string, args []interface{}) {
+	// Write a trailing newline.
+	defer writer.Write([]byte("\n"))
+
+	// Tag is used during unit tests when we might have multiple Rafts all
+	// outputting to the same log.
+	if v.tag != "" {
+		writer.Write([]byte(fmt.Sprintf("(%s) ", v.tag)))
+	}
+
+	// This is a little jank because the depth depends on the internals of
+	// the logxi library, but it's only used during unit tests.
+	if v.withLine {
+		_, file, line, ok := runtime.Caller(4)
+		if !ok {
+			file = "???"
+			line = 0
+		}
+		writer.Write([]byte(fmt.Sprintf("%s:%d ", path.Base(file), line)))
+	}
+
+	writer.Write([]byte(time.Now().Local().Format("2006/01/02 15:04:05.000000")))
+
+	switch level {
+	case log.LevelCritical:
+		writer.Write([]byte(" [CRIT ] "))
+	case log.LevelError:
+		writer.Write([]byte(" [ERROR] "))
+	case log.LevelWarn:
+		writer.Write([]byte(" [WARN ] "))
+	case log.LevelInfo:
+		writer.Write([]byte(" [INFO ] "))
```

Comment:
I don't want to get into a flamewar, but I find this a bit jarring. Here's an example from the unit tests:
```
(server-5a6f) peer.go:607 2016/10/28 11:26:01.944915 [INFO ] Starting heartbeat RPC for peer: id=server-e71d address=e71d
(server-5a6f) peer.go:1042 2016/10/28 11:26:01.944985 [INFO ] Not updating next index for peer, already set: id=server-8f3b address=8f3b next_index=3
```

First, the beginning of "[INFO" doesn't line up. Then, even if it did, I think it'd look nicer as [INFO] followed by the necessary spacing after the close bracket.

And another nit, I'd prefer the date on the far left, not the filename. That makes it easier to sort by time if you're catting files together or something.